packages/doenetml-worker-rust/lib-doenetml-core/src/core/state/types/math_expr.rs

```diff
@@ -0,0 +1,429 @@
+    ///
+    /// assert_eq!(
+    ///     expr_with_blanks.to_latex(ToLatexParams::default()),
+    ///     "x + \u{FF3F}"
+    /// );
+    ///
+    /// let hide_blanks = ToLatexParams {
+    ///     show_blanks: false,
+    ///     ..Default::default()
+    /// };
+    /// assert_eq!(expr_with_blanks.to_latex(hide_blanks), "x + ");
+    /// ```
+    pub fn to_latex(&self, params: ToLatexParams) -> String {
+        match math_to_latex(&self.math_object, params) {
+            Ok(res) => res,
+            Err(_) => BLANK_MATH_OBJECT.to_owned(),
+        }
+    }
+
+    /// Create a new mathematical expression formed by substituting variables with new expressions
+    ///
+    /// Parameters:
+    /// - `substitutions`: a `HashMap` mapping variables to new expressions
+    ///
+    /// Example:
+    ///
+    /// ```no_run
+    /// # use doenetml_core::state::types::math_expr::{MathExpr, MathOrPrimitive, ToLatexParams};
+    /// # use std::collections::HashMap;
+    /// let expr1 = MathExpr::from_text("x+y", true, &["f"]);
+    ///
+    /// let substitutions =
+    ///     HashMap::from([("y".to_string(), MathOrPrimitive::VariableName("z".to_string()))]);
+    /// let expr2 = expr1.substitute(&substitutions);
+    ///
+    /// assert_eq!(expr2.to_latex(ToLatexParams::default()), "x + z");
+    /// ```
+    pub fn substitute(&self, substitutions: &HashMap<String, MathOrPrimitive>) -> MathExpr {
+        let math_object = match substitute_into_math(&self.math_object, substitutions) {
+            Ok(res) => res,
+            Err(_) => BLANK_MATH_OBJECT.to_owned(),
+        };
+
+        MathExpr { math_object }
+    }
+
+    /// Normalize a mathematical expression as specified in `params`.
+    ///
+    /// Example:
+    ///
+    /// ```no_run
+    /// # use doenetml_core::state::types::math_expr::{MathExpr, MathSimplify, NormalizeParams, ToLatexParams};
+    /// let expr = MathExpr::from_text("(x+x+1)(2y+1-y)", true, &["f"]);
+    ///
+    /// let lp = ToLatexParams::default();
+    ///
+    /// let simplify_expand = NormalizeParams {
+    ///     simplify: MathSimplify::Full,
+    ///     expand: true,
+    ///     ..Default::default()
+    /// };
+    ///
+    /// assert_eq!(
+    ///     expr.normalize(simplify_expand).to_latex(lp),
+    ///     "2 x y + 2 x + y + 1"
+    /// );
+    /// ```
+    pub fn normalize(&self, params: NormalizeParams) -> MathExpr {
+        let math_object = match normalize_math(&self.math_object, params) {
+            Ok(res) => res,
+            Err(_) => BLANK_MATH_OBJECT.to_owned(),
+        };
+
+        MathExpr { math_object }
+    }
+
+    /// Create a new mathematical expression by adding `term` to the current expression.
+    pub fn add(&self, term: MathOrPrimitive) -> MathExpr {
+        let term = serde_json::to_string(&term).unwrap();
+
+        let js_source = format!(
+            r#"JSON.parse('{}', serializedComponentsReviver).add(JSON.parse('{}', serializedComponentsReviver));"#,
```

Comment:
If someone uses a `'` in a math expression, does that break this component? This type of code makes me nervous about code injections. We should have a `sanatize` function to make sure that we're not going to break things with this string manipulation.

Reply:
I don't think the ast of math-expressions uses `'`, but one could put a `'` inside a `MathOrPrimitive::VariableName`, which will break this. But, `VariableName` isn't supposed to handle that case. (I'm not sure what to call it. Maybe I'll call it `Symbol`.)

I'm not comfortable with any of these `eval_js` uses, but I got tired of making an extra Javascript function for each of these, so put these in for now. My plan is to eventually remove them and create separate Javascript functions like I did for `normalize`, etc.

I want to get a better handle on what api we actually need to create on `MathExpr` and then come up with the best approach to communicating between Rust and JS. Does what I did for `normalize_math`, etc., seem reasonable and should be reproduced for other functions like these? Or is there a better way?

Reply:
redid this